### ISSUE TITLE
Gradient Accumulation

### ISSUE THREAD

**Issue by mehdizemni** [ultralytics/yolov5#1630]
## ❔Question
@glenn-jocher 
I have a question regarding  **Gradient Accumulation** implemented in train.py.
It seems to me that gradients are accumulated and summed each accumulate step.
The accumulation step is defined like this:

```python
nbs = 64  # nominal batch size
accumulate = max(round(nbs / total_batch_size), 1)  # accumulate loss before optimizing
```
And the optimizer step is not called only after backward steps defined by **accumulate**.
```python
# Backward
scaler.scale(loss).backward()

# Optimize
if ni % accumulate == 0:
    scaler.step(optimizer)  # optimizer.step
    scaler.update()
    optimizer.zero_grad()
    if ema:
        ema.update(model)
```
So if we set any batch-size smaller than 64 we will have the same result as if we were using 64?

**Comment by glenn-jocher** [comment]
@mehdizemni yes that is correct. This allows for training at smaller batch-sizes with similar effect to training at larger batch sizes, up to 64. Past batch-size 64 the accumulate value stays at 1, so every batch is used to optimize.

Note that this does not produce identical results however, because the nn.Batchnorm2d() layers are still averaged every step, there is no accumulation there. So for example training at smaller batch sizes will tend to produce worse results in fully trained models.
